MPTCP: add support and listen option

mux99 opened the pull request:
This PR is internal, and aims to implement MPTCP (defined in RFC8684) into the nginx socket creation process.
The current approach is to add an option to the listen directive (namely mptcp) to enable and disable the protocol.

It would be interesting to have it set by the httpor server blocks but I don't currently see a good way of doing so.

In this initial commit MPTCP is disabled by default simply because it is easy to implement, but it is preferable to have is on by default and give the option to user to explicitly disable it.

matttbe commented on src/core/ngx_connection.c:
```diff
@@ -78,6 +78,7 @@ ngx_create_listening(ngx_conf_t *cf, struct sockaddr *sockaddr,
 
     ls->fd = (ngx_socket_t) -1;
     ls->type = SOCK_STREAM;
+    ls->protocol = 0;
```
I guess this is not needed, because they did that before: `ngx_memzero(ls, sizeof(ngx_listening_t));`

matttbe commented on src/core/ngx_connection.c:
```diff
@@ -487,8 +488,16 @@ ngx_open_listening_sockets(ngx_cycle_t *cycle)
+            s = (ngx_socket_t) -1;
+            if (ls[i].protocol > 0) {
+                s = ngx_socket(ls[i].sockaddr->sa_family, ls[i].type,
+                               ls[i].protocol);
```
probably good to add a comment here:

    /* In case of error, retry with the default protocol */

matttbe commented on src/http/ngx_http_core_module.c:
```diff
@@ -4049,6 +4049,13 @@ ngx_http_core_listen(ngx_conf_t *cf, ngx_command_t *cmd, void *conf)
+            lsopt.protocol = IPPROTO_MPTCP;
+            continue;
+        }
+#endif
```
I guess you also need a modification not to allow using QUIC with MPTCP, see line 4319

matttbe commented on src/core/ngx_connection.c:
```diff
@@ -78,6 +78,7 @@ ngx_create_listening(ngx_conf_t *cf, struct sockaddr *sockaddr,
 
```
(not related to this part of the code) For the commit message: the limit of chars per line is ~72, not 80. (because 'git show' indents it I think)

Also, the links (`[1]`, etc.) are missing, see your lighttpd commit.

Maybe this could be split into 4 commits, not sure what they usually do (check the git history): one for the core, then one for each feature: http, mail, stream.

Then the titles would be: `core|http|mail|stream: add MPTCP support`

matttbe commented on src/core/ngx_connection.c:
```diff
@@ -78,6 +78,7 @@ ngx_create_listening(ngx_conf_t *cf, struct sockaddr *sockaddr,
 
     ls->fd = (ngx_socket_t) -1;
```
(not related to this part of the code) no doc and/or config files to change?

mux99 replied:
the config included as the default one does not contain any mention of more specialized options such as quic or fastopen. I believe then it would be appropriate to not mention MPTCP in the default config but add it to the following pages of the wiki:
- [http](http://nginx.org/en/docs/http/ngx_http_core_module.html#listen)
- [mail](http://nginx.org/en/docs/mail/ngx_mail_core_module.html#listen)
- [stream](http://nginx.org/en/docs/stream/ngx_stream_core_module.html#listen)

which should be submitted as a pull-request on the [github](https://github.com/nginx/nginx.org) for the wiki

matttbe replied:
OK, strange, for example they only added `fastopen` two weeks ago: https://github.com/nginx/nginx.org/commit/30e778f0516f06f6ba7022834112d446cdf50aba

Maybe you can at least modify the VIM plugin: https://github.com/nginx/nginx/blob/bf3e6538b93375e8ac49d3863f74ee7cbe588111/contrib/vim/syntax/nginx.vim#L69

matttbe commented on src/core/ngx_connection.c:
```diff
@@ -78,6 +78,7 @@ ngx_create_listening(ngx_conf_t *cf, struct sockaddr *sockaddr,
 
     ls->fd = (ngx_socket_t) -1;
     ls->type = SOCK_STREAM;
```
(not related to this part of the code) no test to adapt?

mux99 replied:
Tests are hosted [here](http://hg.nginx.org/nginx-tests) as a mercurial repo, but the documentation does not mention the addition of new tests. I would assume that we are not expected to add any to the suite.

matttbe reviewed (changes requested):
Thanks! Looks good!

I guess you still have to update the doc, the config, and the tests ;)

matttbe commented on src/core/ngx_connection.c:
```diff
@@ -487,8 +487,17 @@ ngx_open_listening_sockets(ngx_cycle_t *cycle)
                 continue;
```
For the commit message of `Core: added socket protocol`:

- usually, we use the present form, but looking at the history here, they seem to do like you, except that they always add a `.` at the end (no idea why, but best to do the same)
- Start your sentences with a capital letter: `structure. the`
- What is very important in a commit message is to add the reason why this commit is there: we can easily see what you modified, understand why you did that is harder. You can add something:
```
Modules will then be able to specify a different protocol, e.g.
IPPROTO_MPTCP.
```

matttbe commented on src/core/ngx_connection.c:
```diff
@@ -487,8 +487,17 @@ ngx_open_listening_sockets(ngx_cycle_t *cycle)
+                               ls[i].protocol);
+            }
 
+            /* In case of error, retry with the default protocol */
```
I think it is best to move it in the block above, after having called `ngx_socket`, no?
It is only in this case that we will retry in case of error.

(or we log the error (not critical) and mention we will retry with the default protocol?)

matttbe commented on src/http/ngx_http_core_module.c:
```diff
@@ -4049,6 +4049,13 @@ ngx_http_core_listen(ngx_conf_t *cf, ngx_command_t *cmd, void *conf)
 
+#if (NGX_LINUX)
+        if (ngx_strcmp(value[n].data, "mptcp") == 0) {
+            lsopt.protocol = IPPROTO_MPTCP;
```
mmh, I guess we will need:

```
#ifndef IPPROTO_MPTCP
#define IPPROTO_MPTCP 262
#endif
```

not to have compilation errors on older platforms. But not sure where it would be best to do that.

mux99 replied:
I believe it is required in each of the modules (`http`, `mail`, `stream`) as they can be compiled independently of each other.

matttbe replied:
Yes, but this could be set in a common header (I didn't really check, maybe something like `ngx_socket.h`)?

matttbe replied:
> Yes, but this could be set in a common header (I didn't really check, maybe something like `ngx_socket.h`)?

(was it not possible to do ↑?)

matttbe commented on src/http/ngx_http.c:
```diff
@@ -1845,6 +1845,7 @@ ngx_http_add_listening(ngx_conf_t *cf, ngx_http_conf_addr_t *addr)
 #endif
```
commit: I think you should change:

> This patch and is sister commits add the `mptcp` parameter to the `listen` directive.

and have it specific per commit:

    This patch adds a new parameter 'mptcp' to the  'listen' directive
    in the HTTP module.

Also, you can do the same modification for the titles as mentioned in your first commit:

     HTTP: added MPTCP support.

And I think you can add something like:

    Even if MPTCP is supported by different OS, only Linux supports the
    `IPPROTO_MPTCP` protocol, which is why this feature is currently
    limited to Linux only.


Same in the two next commits of course.

matttbe commented on src/mail/ngx_mail_core_module.c:
```diff
@@ -468,6 +468,11 @@ ngx_mail_core_listen(ngx_conf_t *cf, ngx_command_t *cmd, void *conf)
 #endif
         }
 
+        if (ngx_strcmp(value[i].data, "mptcp") == 0) {
```
out of curiosity: why did you put this block here and not at the end? Same in the other modules. I don't know if there is an order (alphabetical, when they are added, etc.)

mux99 replied:
I very well could have, but it seemed more readable to me since the next parameter (so_keepalive) is quite long.
If it is a bother, it can be moved.

matttbe replied:
what is important is to keep the logic already in place  (alphabetical, when they are added, etc.). If it is random, then fine

matttbe reviewed (changes requested):
thx! Almost there

Commit:
Core: added socket protocol.

This patch updates the creation of listening sockets to use a new field
of the `ngx_listening_s` structure. The `protocol` field can be used in
conjunction with the `type` to specify the protocol to be used.

Modules will then be able to specify a different protocol, e.g.
IPPROTO_MPTCP.

Commit:
HTTP: added MPTCP support.

Multipath TCP (MPTCP), standardized in RFC8684 [1], is a TCP extension
that enables a TCP connection to use different paths.

Multipath TCP has been used for several use cases. On smartphones, MPTCP
enables seamless handovers between cellular and Wi-Fi networks while
preserving established connections. This use-case is what pushed Apple
to use MPTCP since 2013 in multiple applications [2]. On dual-stack
hosts, Multipath TCP enables the TCP connection to automatically use the
best performing path, either IPv4 or IPv6. If one path fails, MPTCP
automatically uses the other path.

To benefit from MPTCP, both the client and the server have to support
it. Multipath TCP is a backward-compatible TCP extension that is enabled
by default on recent Linux distributions (Debian, Ubuntu, Redhat, ...).
Multipath TCP is included in the Linux kernel since version 5.6 [3].
To use it on Linux, an application must explicitly enable it when
creating the socket. No need to change anything else in the application.

Even if MPTCP is supported by different OS, only Linux supports the
`IPPROTO_MPTCP` protocol, which is why this feature is currently
limited to Linux only.

This patch adds a new parameter 'mptcp' to the  'listen' directive
in the HTTP module.

Link: https://www.rfc-editor.org/rfc/rfc8684.html [1]
Link: https://www.tessares.net/apples-mptcp-story-so-far/ [2]
Link: https://www.mptcp.dev [3]

Commit:
Stream: added MPTCP support.

Multipath TCP (MPTCP), standardized in RFC8684 [1], is a TCP extension
that enables a TCP connection to use different paths.

Multipath TCP has been used for several use cases. On smartphones, MPTCP
enables seamless handovers between cellular and Wi-Fi networks while
preserving established connections. This use-case is what pushed Apple
to use MPTCP since 2013 in multiple applications [2]. On dual-stack
hosts, Multipath TCP enables the TCP connection to automatically use the
best performing path, either IPv4 or IPv6. If one path fails, MPTCP
automatically uses the other path.

To benefit from MPTCP, both the client and the server have to support
it. Multipath TCP is a backward-compatible TCP extension that is enabled
by default on recent Linux distributions (Debian, Ubuntu, Redhat, ...).
Multipath TCP is included in the Linux kernel since version 5.6 [3].
To use it on Linux, an application must explicitly enable it when
creating the socket. No need to change anything else in the application.

Even if MPTCP is supported by different OS, only Linux supports the
`IPPROTO_MPTCP` protocol, which is why this feature is currently
limited to Linux only.

This patch adds a new parameter 'mptcp' to the  'listen' directive
in the Stream module.

Link: https://www.rfc-editor.org/rfc/rfc8684.html [1]
Link: https://www.tessares.net/apples-mptcp-story-so-far/ [2]
Link: https://www.mptcp.dev [3]

Commit:
Mail: added MPTCP support.

Multipath TCP (MPTCP), standardized in RFC8684 [1], is a TCP extension
that enables a TCP connection to use different paths.

Multipath TCP has been used for several use cases. On smartphones, MPTCP
enables seamless handovers between cellular and Wi-Fi networks while
preserving established connections. This use-case is what pushed Apple
to use MPTCP since 2013 in multiple applications [2]. On dual-stack
hosts, Multipath TCP enables the TCP connection to automatically use the
best performing path, either IPv4 or IPv6. If one path fails, MPTCP
automatically uses the other path.

To benefit from MPTCP, both the client and the server have to support
it. Multipath TCP is a backward-compatible TCP extension that is enabled
by default on recent Linux distributions (Debian, Ubuntu, Redhat, ...).
Multipath TCP is included in the Linux kernel since version 5.6 [3].
To use it on Linux, an application must explicitly enable it when
creating the socket. No need to change anything else in the application.

Even if MPTCP is supported by different OS, only Linux supports the
`IPPROTO_MPTCP` protocol, which is why this feature is currently
limited to Linux only.

This patch adds a new parameter 'mptcp' to the  'listen' directive
in the Mail module.

Link: https://www.rfc-editor.org/rfc/rfc8684.html [1]
Link: https://www.tessares.net/apples-mptcp-story-so-far/ [2]
Link: https://www.mptcp.dev [3]

matttbe commented on src/core/ngx_connection.c:
```diff
@@ -487,8 +487,19 @@ ngx_open_listening_sockets(ngx_cycle_t *cycle)
+                s = ngx_socket(ls[i].sockaddr->sa_family, ls[i].type,
+                               ls[i].protocol);
+                /* In case of error, retry with the default protocol */
+                ngx_log_error(NGX_LOG_NOTICE, log, 0,
```
It looks like you will always print something, no matter if there was an error or not, no?
